File: src/main/java/fi/metatavu/oioi/cm/rest/translate/WallResourceTranslator.java
package fi.metatavu.oioi.cm.rest.translate;

import java.util.Map;
import java.util.stream.Collectors;

import javax.inject.Inject;

import fi.metatavu.oioi.cm.persistence.model.Resource;
import fi.metatavu.oioi.cm.persistence.model.ResourceProperty;
import fi.metatavu.oioi.cm.persistence.model.ResourceStyle;
import fi.metatavu.oioi.cm.resources.ResourceController;
import fi.metatavu.oioi.cm.wall.WallResource;

/**
 * Translator for WallResource 
 * 
 * @author Antti Leppä
 * @author Heikki Kurhinen
 */
public class WallResourceTranslator extends AbstractTranslator<fi.metatavu.oioi.cm.persistence.model.Resource, WallResource> {
  
  @Inject
  private ResourceController resourceController;

  @Override
  public WallResource translate(fi.metatavu.oioi.cm.persistence.model.Resource entity) {
    if (entity == null) {
      return null;
    }

    WallResource result = new WallResource();
    result.setSlug(entity.getSlug());
    result.setChildren(translate(resourceController.listResourcesByParent(entity)));
    result.setData(entity.getData());
    result.setName(entity.getName());
    result.setProperties(getProperties(entity));
    result.setStyles(getStyles(entity));
    result.setType(entity.getType());
    
    return result;
  }

  /**
   * Translates styles
   * 
   * @param entity resource
   * @return styles as key value pairs
   */
  private Map<String, String> getStyles(Resource entity) {
    return resourceController.listStyles(entity).stream().collect(Collectors.toMap(ResourceStyle::getKey, ResourceStyle::getValue));
  }

  /**
   * Translates styles
   * 
   * @param entity resource
   * @return styles as key value pairs
   */
  private Map<String, String> getProperties(Resource entity) {
    return resourceController.listProperties(entity).stream().collect(Collectors.toMap(ResourceProperty::getKey, ResourceProperty::getValue));
  }
  
}

File: src/main/kotlin/fi/metatavu/oioi/cm/rest/translate/WallResourceTranslator.kt
package fi.metatavu.oioi.cm.rest.translate

import fi.metatavu.oioi.cm.model.WallResource
import fi.metatavu.oioi.cm.persistence.model.Resource
import fi.metatavu.oioi.cm.resources.ResourceController
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Translator for WallResource
 *
 * @author Antti Leppä
 * @author Heikki Kurhinen
 */
@ApplicationScoped
class WallResourceTranslator : AbstractTranslator<Resource?, WallResource?>() {

    @Inject
    lateinit var resourceController: ResourceController

    override fun translate(entity: Resource?): WallResource? {
        return translate(entity = entity, applicationName = entity?.name)
    }

    /**
     * Translates wall resource
     *
     * @param entity resource entity
     * @param applicationName application name
     * @return translated wall resource
     */
    fun translate(entity: Resource?, applicationName: String?): WallResource? {
        entity ?: return null
        applicationName ?: return null

        val resourceStyles = resourceController.listStyles(entity)
        val resourceProperties = resourceController.listProperties(entity)

        val modifiedAt = resourceStyles.mapNotNull { it.modifiedAt }
            .plus(resourceProperties.mapNotNull{ it.modifiedAt })
            .plus(entity.modifiedAt!!)
            .maxOf { it }

        val children = translate(resourceController.listResourcesByParent(parent = entity, resourceType = null))
            .mapNotNull { it }

        return WallResource(
            slug = entity.slug!!,
            children = children,
            data = entity.data,
            name = applicationName,
            properties = resourceProperties.associate { it.key!! to it.value!! },
            styles = resourceStyles.associate { it.key!! to it.value!! },
            type = entity.type!!,
            modifiedAt = modifiedAt
        )
    }

}